ls1intum/Artemis
# Allow instructors to add results for external submissions

### Checklist

### Motivation and Context
This PR addresses issue #1121. The use case is the following: Sometimes students submit their work off the Artemis platform, e.g. via email. Right now, it is not possible to feed these submissions into Artemis so that they can be assessed together with the regular submissions. This PR is the first step to solving this problem. It provides instructors with a way to create manual results for an exercise by providing the student login and the result details.

In the background, a participation is created (if it does not exist yet), an empty submission is created and a result is created according to the details that the instructor entered into the creation dialog. The dialog can be accessed by instructors from the exercise detail page via a button `Add external submission`. The aforementioned applies to all exercise types but Quiz exercises.

### Description
- Add new components `external-submission-button` and `external-submission-dialog`
- Insert new button `Add External Submission` onto the administrative detail pages of Programming, Modeling, Text, File Upload exercises
- Add service `external-submission-service` that calls backend endpoint `/exercises/{exerciseId}/external-submission-results` to create a new participation, submission and result (with the student login as a query param and the result to be created as the request body)
- Add new function `createParticipationWithEmptySubmission` to ParticipationService that creates a finished participation for the given student and exercise (or uses the existing participation) and then initialized a submission with the new submission type `EXTERNAL` and sets the status to `submitted`
- The backend endpoint calls this function and then creates a result for the created submission. If a result already exists for the submission, no changes are performed and an alert "A result already exists for this submission" is displayed

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Click on Exercises for a course
4. Click on View for an exercise
5. Click on the yelow button `Add External Submission`
6. Fill out the form and click Save
7. Check that a participation, submission and result were created correctly

### Screenshots
![OnPaste 20200124-022745](https://user-images.githubusercontent.com/7109225/73038124-9289d100-3e51-11ea-9989-5bbaf16e681a.png)
![OnPaste 20200124-022819](https://user-images.githubusercontent.com/7109225/73038125-9289d100-3e51-11ea-8c6e-3c1921f75489.png)
